### ISSUE TITLE
Slow queries when hitting the cache

### ISSUE THREAD

**Issue by cedvdb** [firebase/firebase-js-sdk#8681]
### Operating System

ubuntu

### Environment (if applicable)

Version 131.0.6778.85 (Official Build) (64-bit)

### Firebase SDK Version

10.11.1

### Firebase SDK Product(s)

Firestore

### Project Tooling

Flutter

### Detailed Problem Description

I have a minimal reproduction sample where a simple query hitting the cache takes a whopping 45 seconds to resolve.

During my trials to reproduce this issue I've found that:

 - The same query, when `disableNetwork()` is called before writing the data to Firestore takes milliseconds. After glancing over index_db, I see that the way data is stored when the network is disabled is as mutation, which is different from when the network is enabled.
 - The same query also resolve normally when there is no flutter animation on the screen.
 
As I could not isolate the issue to any of those technologies, this leads me to believe there are performances issues in both firestore and flutter animation mechanism acting on each other. Furthermore the fact that the query is resolving in milliseconds depending on how it has been written is indicative that there is an issue on the firestore side there.

 
 
 

### Steps and code to reproduce issue

Prerequisite: Flutter is required to launch this 

1. Clone `https://github.com/cedvdb/flutter_animation_perf/tree/circular_progress_indicator_firestore_perf_issue`  branch `circular_progress_indicator_firestore_perf_issue` 
2. Link a firebase project as described here https://firebase.google.com/docs/flutter/setup?platform=web
3. Run the example app on the web
4. Press create data button and wait until the text tells you to reload
5. Reload the page
6. Wait at least 2 minutes for the data to load

![image](https://github.com/user-attachments/assets/f9eef8fa-bbef-4a17-8809-5b3977f101e9)

1000 documents loaded in 41 seconds here.

- After that we are going to delete the local cache, by going in the chrome devtools > application > indexedDB > and then delete all firestore databases.
- Close the application
- switch to the branch `network_disabled`
- relaunch the application
- Press create data button and wait 30 seconds, then reload the page
- It now takes approximately 3 seconds to load the data from cache (still a lot but an order of magnitude less than before).

**Comment by wu-hui** [comment]
What you saw is an expected behavior. When network is disabled, the SDK reads the results from cache without requesting anything from server, hence the faster turn around because there is no network latency involved.

You can achieve the same result without disabling network, by supplying a source option in https://github.com/firebase/flutterfire/blob/9eac89ebf1faa16966733bc527fb17c66efac654/packages/cloud_firestore/cloud_firestore_platform_interface/lib/src/get_options.dart#L32. Setting source to cache directly should work the same way.

Alternatively, use `snapshots` instead of `get`, `snapshots` will give you the results from cache first, then augment the results with server response in subsequent snapshots, and is recommended over `get` in general.

**Comment by cedvdb** [comment]
> hence the faster turn around because there is no network latency involved.

You are misunderstanding the issue.

The faster turnaround being 50 seconds when no network (hitting the cache) request and 1-2 seconds with a network request

**Comment by MarkDuckworth** [comment]
I was reading through this out of curiosity. @cedvdb, your observation is accurate, that the data written to the SDK cache is treated differently between the two tested scenarios. This is intentional. We don't generally get into the details of the implementation, but at a high-level, the data is different types. In the first case, the data is something that has not been written to or acknowledged by the server. In the second case, the cached data is something retrieved from the server. It's not clear to me why you are seeing retrieval of one be more performant than the other.

Re: animation. I'm not too familiar with the flutter animation you are using, but based on the thread you posted on the flutter repo, it sounds like it could be a blocking. Others have reported cpu spikes when using it.

Thanks for posting this and following up with your solution. It may help others resolve the same situation.

**Comment by wu-hui** [comment]
Ah, you are right. Sorry to mis-interpret you. As Mark said, this is a separate issue, queuing up a large number of pending mutations locally can affect query performance.
